controllers/topology/internal/extensions/patches/patch.go

```diff
@@ -0,0 +1,198 @@
+// and then returns itself (for convenient chaining).
+func (o *PatchOptions) ApplyOptions(opts []PatchOption) *PatchOptions {
+	for _, opt := range opts {
+		opt.ApplyToHelper(o)
+	}
+	return o
+}
+
+// PreserveFields instructs the patch func to preserve fields.
+type PreserveFields []contract.Path
+
+// ApplyToHelper applies this configuration to the given patch options.
+func (i PreserveFields) ApplyToHelper(opts *PatchOptions) {
+	opts.preserveFields = i
+}
+
+// patchObject overwrites spec in object with spec.template.spec of patchedTemplate,
+// while preserving the configured fields.
+func patchObject(ctx context.Context, object, patchedTemplate *unstructured.Unstructured, opts ...PatchOption) error {
+	log := tlog.LoggerFrom(ctx)
+
+	patchOptions := &PatchOptions{}
+	patchOptions = patchOptions.ApplyOptions(opts)
+
+	// Create a copy of the object.
+	original := object.DeepCopy()
+
+	if err := copySpec(copySpecInput{
+		src:              patchedTemplate,
+		dest:             object,
+		srcSpecPath:      "spec.template.spec",
+		destSpecPath:     "spec",
+		fieldsToPreserve: patchOptions.preserveFields,
+	}); err != nil {
+		return errors.Wrapf(err, "failed to patch object %s by appying changes from the patched template",
+			tlog.KObj{Obj: object})
+	}
+
+	// Log the delta between the object before and after applying the accumulated patches.
+	helper, err := mergepatch.NewHelper(original, object, nil)
+	if err != nil {
+		return err
+	}
+	log.V(4).WithObject(object).Infof("Applying accumulated patches", "delta", string(helper.Changes()))
+
+	return nil
+}
+
+// patchTemplate overwrites spec.template.spec in template with spec.template.spec of patchedTemplate,
+// while preserving the configured fields.
+func patchTemplate(ctx context.Context, template, patchedTemplate *unstructured.Unstructured, opts ...PatchOption) error {
+	log := tlog.LoggerFrom(ctx)
+
+	patchOptions := &PatchOptions{}
+	patchOptions = patchOptions.ApplyOptions(opts)
+
+	// Create a copy of the template.
+	original := template.DeepCopy()
+
+	if err := copySpec(copySpecInput{
+		src:              patchedTemplate,
+		dest:             template,
+		srcSpecPath:      "spec.template.spec",
+		destSpecPath:     "spec.template.spec",
+		fieldsToPreserve: patchOptions.preserveFields,
+	}); err != nil {
+		return errors.Wrapf(err, "failed to patch template %s by appying changes from the patched template",
+			tlog.KObj{Obj: template})
+	}
+
+	// Log the delta between the object before and after applying the accumulated patches.
+	helper, err := mergepatch.NewHelper(original, template, nil)
+	if err != nil {
+		return err
+	}
+	log.V(4).WithObject(template).Infof("Applying accumulated patches", "delta", string(helper.Changes()))
+
+	return nil
+}
+
+// patchTemplateSpec overwrites spec in templateJSON with spec of patchedTemplateBytes.
+func patchTemplateSpec(templateJSON *apiextensionsv1.JSON, patchedTemplateBytes []byte) error {
```

Comment:
Wondering if we can drop this func and use patchTemplate.
They seems functionally equivalent (the only part of spec we care about spec.template.spec), so the only problem to be addressed is the difference of type, so we need a wrapper, something like patchRequestTemplate

Reply:
It's more than that:
* patchTemplate patches: .spec.template.spec => spec.template.spec. 
* patchTemplateSpec patches: .spec => .spec

Except we want to change the semantic and only want to allow patches to modify `.spec.template.spec` instead of the entire `.spec`. The only other currently existing .spec  field in templates I can think of is `.spec.template.metadata`.

Reply:
> The only other currently existing .spec field in templates I can think of is .spec.template.metadata.

yep, that's my point as well. this is why I think we should dedup

Reply:
The 99% use case is definitely `.spec.template.spec`. We can also start with that and consider expanding on that it if there is demand

Reply:
I refactored patchTemplate/patchObject a bit to deduplicate code and to have a correct diff. I don't think we can reuse patchTemplate or patchUnstructured here as we don't need the same diff logic and logging.

Independent of that we can of course still decide to reduce the scope of potential changes to .spec.template.spec